mopidy/core/actor.py

```diff
@@ -136,6 +140,89 @@ def tags_changed(self, tags):
+                    coverage = ['tracklist', 'mode', 'play-last', 'volume',
+                                'history']
+            if len(coverage):
+                self._load_state(coverage)
+        except Exception as e:
+            logger.warn('Restore state: Unexpected error: %s', str(e))
+
+    def teardown(self):
+        """ Do not call this function. It is for internal use at shutdown."""
+        try:
+            if self._config and 'restore_state' in self._config['core']:
+                if self._config['core']['restore_state']:
+                    self._save_state()
+        except Exception as e:
+            logger.warn('Unexpected error while saving state: %s', str(e))
+
+    def _get_data_dir(self):
+        # get or create data director for core
+        data_dir_path = bytes(
+            os.path.join(self._config['core']['data_dir'], 'core'))
+        path.get_or_create_dir(data_dir_path)
+        return data_dir_path
+
+    def _save_state(self):
+        """
+        Save current state to disk.
+        """
+
+        file_name = os.path.join(self._get_data_dir(), 'persistent.state')
+        logger.info('Save state to %s', file_name)
+
+        data = {}
+        data['version'] = mopidy.__version__
+        data['state'] = CoreState(
+            tracklist=self.tracklist._export_state(),
+            history=self.history._export_state(),
+            playback=self.playback._export_state(),
+            mixer=self.mixer._export_state())
+        storage.save(file_name, data)
+        logger.debug('Save state done.')
+
+    def _load_state(self, coverage):
+        """
+        Restore state from disk.
+
+        Load state from disk and restore it. Parameter `coverage`
+        limits the amount data to restore. Possible
+        values for `coverage` (list of one or more of):
+
+            - 'tracklist' fill the tracklist
+            - 'mode' set tracklist properties (consume, random, repeat, single)
+            - 'autoplay' start playing ('tracklist' also required)
+            - 'volume' set mixer volume
+            - 'history' restore history
+
+        :param coverage: amount of data to restore
+        :type coverage: list of string (see above)
+        """
+
+        file_name = os.path.join(self._get_data_dir(), 'persistent.state')
+        logger.info('Load state from %s', file_name)
+
+        data = storage.load(file_name)
+
+        # Try only once. If something goes wrong, the next start is clean.
+        os.remove(file_name)
+
+        if 'state' in data:
+            core_state = data['state']
+            validation.check_instance(core_state, CoreState)
+            self.history._restore_state(core_state.history, coverage)
+            self.tracklist._restore_state(core_state.tracklist, coverage)
+            # playback after tracklist
+            self.playback._restore_state(core_state.playback, coverage)
+            self.mixer._restore_state(core_state.mixer, coverage)
```

Comment:
In case you use hardware mixing and the volume has been turned up when using some other progrem, and it ended up really high, I think it's best to restore the mixer state before the playback state, so we don't start playing before the volume is at an appropriate level.